File: Src/Icm.Core/Collections/SortedCollection/StructKeyStructValue/BaseSortedCollection.vb
Namespace Icm.Collections.Generic.StructKeyStructValue
    Public MustInherit Class BaseSortedCollection(Of TKey As {Structure, IComparable(Of TKey)}, TValue As Structure)
        Implements ISortedCollection(Of TKey, TValue)

        Public Sub New(ByVal otkey As ITotalOrder(Of TKey))
            _totalOrder = otkey
        End Sub

        ''' <summary>
        ''' 
        ''' </summary>
        ''' <remarks></remarks>
        Private Class RangePointIterator
            Implements IEnumerator(Of Tuple(Of TKey, TValue?)), IEnumerable(Of Tuple(Of TKey, TValue?))

            Private f_ As ISortedCollection(Of TKey, TValue)
            Private rangeStart_ As TKey
            Private rangeEnd_ As TKey
            Private current_ As Tuple(Of TKey, TValue?)

            Private idx_ As Date

            Public Sub New(ByVal f As ISortedCollection(Of TKey, TValue), ByVal rangeStart As TKey?, ByVal rangeEnd As TKey?)
                rangeStart_ = f.TotalOrder.LstIfNull(rangeStart)
                rangeEnd_ = f.TotalOrder.GstIfNull(rangeEnd)
                f_ = f
            End Sub

            Public Sub New(ByVal f As ISortedCollection(Of TKey, TValue), ByVal intf As Vector2(Of TKey?))
                MyClass.New(f, intf.Item1, intf.Item2)
            End Sub

            Public ReadOnly Property Current() As Tuple(Of TKey, TValue?) Implements IEnumerator(Of Tuple(Of TKey, TValue?)).Current
                Get
                    If current_ Is Nothing Then
                        Throw New InvalidOperationException("Enumerator has not been reset")
                    End If
                    Return current_
                End Get
            End Property

            Public ReadOnly Property Current1() As Object Implements System.Collections.IEnumerator.Current
                Get
                    If current_ Is Nothing Then
                        Throw New InvalidOperationException("Enumerator has not been reset")
                    End If
                    Return current_
                End Get
            End Property

            Public Function MoveNext() As Boolean Implements System.Collections.IEnumerator.MoveNext
                If current_ Is Nothing Then
                    current_ = Pair(rangeStart_)
                ElseIf current_.Item1.Equals(rangeEnd_) Then
                    current_ = Nothing
                    Return False
                Else
                    Dim sig = f_.Next(current_.Item1)


                    If sig.HasValue Then
                        If sig.Value.CompareTo(rangeEnd_) > 0 Then
                            current_ = Pair(rangeEnd_)
                            Return False
                        Else
                            current_ = Pair(sig.Value)
                        End If
                    Else
                        If current_.Item1.CompareTo(rangeEnd_) < 0 Then
                            current_ = Pair(rangeEnd_)
                        Else
                            ' NO DEBERÍA SUCEDER
                            Throw New Exception
                        End If
                    End If
                End If

                Return True
            End Function

            ReadOnly Property Pair(ByVal key As TKey) As Tuple(Of TKey, TValue?)
                Get
                    If f_.ContainsKey(key) Then
                        Return New Tuple(Of TKey, TValue?)(key, f_(key))
                    Else
                        Return New Tuple(Of TKey, TValue?)(key, Nothing)
                    End If
                End Get
            End Property

            Public Sub Reset() Implements IEnumerator.Reset
                current_ = Nothing
            End Sub

            Private disposedValue As Boolean = False        ' To detect redundant calls

            ' IDisposable
            Protected Overridable Sub Dispose(ByVal disposing As Boolean)
                If Not Me.disposedValue Then
                    If disposing Then
                        ' free other state (managed objects).
                    End If

                    ' free your own state (unmanaged objects).
                    ' set large fields to null.
                End If
                Me.disposedValue = True
            End Sub

#Region " IDisposable Support "
            ' This code added by Visual Basic to correctly implement the disposable pattern.
            Public Sub Dispose() Implements IDisposable.Dispose
                ' Do not change this code.  Put cleanup code in Dispose(ByVal disposing As Boolean) above.
                Dispose(True)
                GC.SuppressFinalize(Me)
            End Sub
#End Region

            Public Function GetEnumerator() As IEnumerator(Of Tuple(Of TKey, TValue?)) Implements IEnumerable(Of Tuple(Of TKey, TValue?)).GetEnumerator
                Return Me
            End Function

            Public Function GetEnumerator1() As IEnumerator Implements System.Collections.IEnumerable.GetEnumerator
                Return Me
            End Function
        End Class


        Private _totalOrder As ITotalOrder(Of TKey)

        ReadOnly Property TotalOrder As ITotalOrder(Of TKey) Implements ISortedCollection(Of TKey, TValue).TotalOrder
            Get
                Return _totalOrder
            End Get
        End Property

        Public MustOverride Sub Add(ByVal key As TKey, ByVal value As TValue) Implements ISortedCollection(Of TKey, TValue).Add

        Public MustOverride Function ContainsKey(ByVal key As TKey) As Boolean Implements ISortedCollection(Of TKey, TValue).ContainsKey

        Public Function GetFreeKey(ByVal desiredKey As TKey) As TKey? Implements ISortedCollection(Of TKey, TValue).GetFreeKey
            Dim fFinal As TKey?
            fFinal = desiredKey
            Do While fFinal.HasValue AndAlso ContainsKey(fFinal.Value)
                If TotalOrder.Compare(fFinal.Value, TotalOrder.Greatest) = 0 Then
                    Return Nothing
                End If
                fFinal = TotalOrder.Next(fFinal.Value)
            Loop
            Return fFinal
        End Function

        Default Public MustOverride Property Item(ByVal key As TKey) As TValue Implements ISortedCollection(Of TKey, TValue).Item

        Public MustOverride Function [Next](ByVal key As TKey) As TKey? Implements ISortedCollection(Of TKey, TValue).Next

        Public MustOverride Function Previous(ByVal key As TKey) As TKey? Implements ISortedCollection(Of TKey, TValue).Previous

        Public Overrides Function ToString() As String Implements ISortedCollection(Of TKey, TValue).ToString
            Return ToString(TotalOrder.Least, TotalOrder.Greatest)
        End Function


        ''' <summary>
        ''' String representation of an interval of the sorted collection.
        ''' For printing values, ToString will be used.
        ''' </summary>
        ''' <param name="f1">Initial key.</param>
        ''' <param name="f2">Final key.</param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Overridable Overloads Function ToString(ByVal f1 As TKey, ByVal f2 As TKey) As String Implements ISortedCollection(Of TKey, TValue).ToString
            Dim result As New System.Text.StringBuilder

            For Each element In PointEnumerable(f1, f2)
                If element.Item2 Is Nothing Then
                    result.AppendFormat("NC {0} ---" & vbCrLf, element.Item1)
                Else
                    result.AppendFormat("-> {0} {1}" & vbCrLf, element.Item1, element.Item2.ToString)
                End If
            Next

            Return result.ToString
        End Function

        Public MustOverride Function KeyOrNext(ByVal key As TKey) As TKey? Implements ISortedCollection(Of TKey, TValue).KeyOrNext

        Public MustOverride Function KeyOrPrev(ByVal key As TKey) As TKey? Implements ISortedCollection(Of TKey, TValue).KeyOrPrev

        Public MustOverride Sub Remove(ByVal key As TKey) Implements ISortedCollection(Of TKey, TValue).Remove

        Public Function IntervalEnumerable(ByVal intf As Vector2(Of TKey?)) As IEnumerable(Of Vector2(Of Tuple(Of TKey, TValue?))) Implements ISortedCollection(Of TKey, TValue).IntervalEnumerable
            Return IntervalEnumerable(intf.Item1, intf.Item2)
        End Function

        Public Function IntervalEnumerable(ByVal intStart As TKey?, ByVal intEnd As TKey?) As IEnumerable(Of Vector2(Of Tuple(Of TKey, TValue?))) Implements ISortedCollection(Of TKey, TValue).IntervalEnumerable
            Dim it = New PairEnumerator(Of Tuple(Of TKey, TValue?))(
                New RangePointIterator(Me, intStart, intEnd))
            Return it
        End Function

        Public MustOverride Function Count() As Integer Implements ISortedCollection(Of TKey, TValue).Count

        Public Function PointEnumerable(ByVal intStart As TKey?, ByVal intEnd As TKey?) As System.Collections.Generic.IEnumerable(Of System.Tuple(Of TKey, TValue?)) Implements ISortedCollection(Of TKey, TValue).PointEnumerable
            Return New RangePointIterator(Me, intStart, intEnd)
        End Function

        Public Function PointEnumerable(ByVal intf As Vector2(Of TKey?)) As System.Collections.Generic.IEnumerable(Of System.Tuple(Of TKey, TValue?)) Implements ISortedCollection(Of TKey, TValue).PointEnumerable
            Return PointEnumerable(intf.Item1, intf.Item2)
        End Function

        Public Function IntervalEnumerable() As System.Collections.Generic.IEnumerable(Of Vector2(Of System.Tuple(Of TKey, TValue?))) Implements ISortedCollection(Of TKey, TValue).IntervalEnumerable
            Return IntervalEnumerable(Nothing, Nothing)
        End Function

        Public Function PointEnumerable() As System.Collections.Generic.IEnumerable(Of System.Tuple(Of TKey, TValue?)) Implements ISortedCollection(Of TKey, TValue).PointEnumerable
            Return PointEnumerable(Nothing, Nothing)
        End Function
    End Class
End Namespace

File: Src/Icm.Core/Collections/Generic/StructKeyStructValue/BaseSortedCollection.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace Icm.Collections.Generic.StructKeyStructValue
{
    public abstract class BaseSortedCollection<TKey, TValue> : ISortedCollection<TKey, TValue> where TKey : struct, IComparable<TKey> where TValue : struct
    {
        protected BaseSortedCollection(ITotalOrder<TKey> otkey)
        {
            TotalOrder = otkey;
        }

        /// <summary>
        /// 
        /// </summary>
        /// <remarks></remarks>
        private sealed class RangePointIterator : IEnumerator<Tuple<TKey, TValue?>>, IEnumerable<Tuple<TKey, TValue?>>
        {

            private readonly ISortedCollection<TKey, TValue> _f;
            private readonly TKey _rangeStart;
            private readonly TKey _rangeEnd;

            private Tuple<TKey, TValue?> _current;
            
            public RangePointIterator(ISortedCollection<TKey, TValue> f, TKey? rangeStart, TKey? rangeEnd)
            {
                _rangeStart = f.TotalOrder.LstIfNull(rangeStart);
                _rangeEnd = f.TotalOrder.GstIfNull(rangeEnd);
                _f = f;
            }
            
            public Tuple<TKey, TValue?> Current
            {
                get
                {
                    if (_current == null)
                    {
                        throw new InvalidOperationException("Enumerator has not been reset");
                    }
                    return _current;
                }
            }

            object IEnumerator.Current
            {
                get
                {
                    if (_current == null)
                    {
                        throw new InvalidOperationException("Enumerator has not been reset");
                    }

                    return _current;
                }
            }

            public bool MoveNext()
            {
                if (_current == null)
                {
                    _current = Pair(_rangeStart);
                }
                else if (_current.Item1.Equals(_rangeEnd))
                {
                    _current = null;
                    return false;
                }
                else
                {
                    var sig = _f.Next(_current.Item1);


                    if (sig.HasValue)
                    {
                        if (sig.Value.CompareTo(_rangeEnd) > 0)
                        {
                            _current = Pair(_rangeEnd);
                            return false;
                        }

                        _current = Pair(sig.Value);
                    }
                    else
                    {
                        if (_current.Item1.CompareTo(_rangeEnd) < 0)
                        {
                            _current = Pair(_rangeEnd);
                        }
                        else
                        {
                            // NO DEBERÍA SUCEDER
                            throw new Exception();
                        }
                    }
                }

                return true;
            }

            private Tuple<TKey, TValue?> Pair(TKey key)
            {
                return _f.ContainsKey(key) 
                    ? new Tuple<TKey, TValue?>(key, _f[key]) 
                    : new Tuple<TKey, TValue?>(key, null);
            }

            public void Reset()
            {
                _current = null;
            }
            
            public void Dispose()
            {

            }

            public IEnumerator<Tuple<TKey, TValue?>> GetEnumerator()
            {
                return this;
            }
            
            IEnumerator IEnumerable.GetEnumerator()
            {
                return this;
            }
        }


        public ITotalOrder<TKey> TotalOrder { get; }

        public abstract void Add(TKey key, TValue value);

        public abstract bool ContainsKey(TKey key);

        public TKey? GetFreeKey(TKey desiredKey)
        {
            TKey? fFinal = desiredKey;
            while (ContainsKey(fFinal.Value))
            {
                if (TotalOrder.Compare(fFinal.Value, TotalOrder.Greatest()) == 0)
                {
                    return null;
                }

                fFinal = TotalOrder.Next(fFinal.Value);
            }

            return fFinal;
        }

        public abstract TValue this[TKey key] { get; set; }

        public abstract TKey? Next(TKey key);

        public abstract TKey? Previous(TKey key);

        public override string ToString()
        {
            return ToString(TotalOrder.Least(), TotalOrder.Greatest());
        }


        /// <summary>
        /// String representation of an interval of the sorted collection.
        /// For printing values, ToString will be used.
        /// </summary>
        /// <param name="f1">Initial key.</param>
        /// <param name="f2">Final key.</param>
        /// <returns></returns>
        /// <remarks></remarks>
        public virtual string ToString(TKey f1, TKey f2)
        {
            System.Text.StringBuilder result = new System.Text.StringBuilder();

            foreach (var element in PointEnumerable(f1, f2))
            {
                if (element.Item2 == null)
                {
                    result.AppendFormat("NC {0} ---\r\n", element.Item1);
                }
                else
                {
                    result.AppendFormat("-> {0} {1}\r\n", element.Item1, element.Item2);
                }
            }

            return result.ToString();
        }

        public abstract TKey? KeyOrNext(TKey key);

        public abstract TKey? KeyOrPrev(TKey key);

        public abstract void Remove(TKey key);

        public IEnumerable<Vector2<Tuple<TKey, TValue?>>> IntervalEnumerable(Vector2<TKey?> intf)
        {
            return IntervalEnumerable(intf.Item1, intf.Item2);
        }

        public IEnumerable<Vector2<Tuple<TKey, TValue?>>> IntervalEnumerable(TKey? intStart, TKey? intEnd)
        {
            return new PairEnumerator<Tuple<TKey, TValue?>>(new RangePointIterator(this, intStart, intEnd));
        }

        public abstract int Count();

        public IEnumerable<Tuple<TKey, TValue?>> PointEnumerable(TKey? intStart, TKey? intEnd)
        {
            return new RangePointIterator(this, intStart, intEnd);
        }

        public IEnumerable<Tuple<TKey, TValue?>> PointEnumerable(Vector2<TKey?> intf)
        {
            return PointEnumerable(intf.Item1, intf.Item2);
        }

        public IEnumerable<Vector2<Tuple<TKey, TValue?>>> IntervalEnumerable()
        {
            return IntervalEnumerable(null, null);
        }

        public IEnumerable<Tuple<TKey, TValue?>> PointEnumerable()
        {
            return PointEnumerable(null, null);
        }
    }
}
